File: goAlgo/qs/qs.go
package main

import (
	"fmt"
	"math/rand/v2"
)

func qs(arr *[]int) {
	length := len(*arr) //17
	if length < 2 {
		return
	}
	mid := part(arr) //16
	l, r := (*arr)[:mid], (*arr)[mid+1:length]
	fmt.Println(l, r)
	qs(&l)
	qs(&r)
}

func part(arrPtr *[]int) int {
	arr, end := *arrPtr, len(*arrPtr)-1
	pivot := rand.IntN(end + 1)
	swap(&arr, pivot, end)
	i := -1
	for j := 0; j < len(arr); j++ {
		if arr[j] >= arr[len(arr)-1] {
			continue
		}
		i++
		swap(&arr, i, j)
	}
	i++
	swap(&arr, i, end)
	return i
}

func swap(arr *[]int, a int, b int) {
	(*arr)[a], (*arr)[b] = (*arr)[b], (*arr)[a]
}

func main() {
	inp := []int{7, 4, 7, 9, 0, 1, 2, 8, 45, 3, 54, 5, 3, 4, 7, 1, 0}
	qs(&inp)
	fmt.Println("final:", inp)
}

File: qs.ts
export function qs(arr: number[], lo: number, hi: number): void {
  if (hi - lo < 1) return;
  const mid = part(arr, lo, hi);
  qs(arr, lo, mid - 1);
  qs(arr, mid + 1, hi);
}

function part(arr: number[], lo: number, hi: number): number {
  const pivotIdx = Math.floor((Math.random() * (hi - lo + 1)) / 2) + lo;
  swap(arr, pivotIdx, hi);
  let i = lo - 1;
  for (let j = lo; j < hi; j++) {
    if (arr[j] >= arr[hi]) continue;
    i++;
    swap(arr, i, j);
  }
  i++;
  swap(arr, i, hi);
  return i;
}

function swap(arr: number[], first: number, second: number) {
  [arr[first], arr[second]] = [arr[second], arr[first]];
}

const inp = [2, 1, 3, 2, 5, 3, 8, 9, 5, 4, 2, 0, 3];

qs(inp, 0, inp.length - 1);
console.log(inp);
